src/collectors/redisstat/redisstat.py

```diff
@@ -69,68 +80,93 @@ class RedisCollector(diamond.collector.Collector):
-            if auth is not None:
-                instance_list.append('%s:%d/%s' % (host, port, auth))
-            else:
-                instance_list.append('%s:%d' % (host, port))
-
-        self.instances = {}
-        for instance in instance_list:
-
-            if '@' in instance:
-                (nickname, hostport) = instance.split('@', 1)
-            else:
-                nickname = None
-                hostport = instance
-
-            if '/' in hostport:
-                parts = hostport.split('/')
-                hostport = parts[0]
-                auth = parts[1]
-            else:
-                auth = None
-
-            if ':' in hostport:
-                if hostport[0] == ':':
-                    host = self._DEFAULT_HOST
-                    port = int(hostport[1:])
-                else:
-                    parts = hostport.split(':')
-                    host = parts[0]
-                    port = int(parts[1])
-            else:
-                host = hostport
-                port = self._DEFAULT_PORT
-
-            if nickname is None:
-                nickname = str(port)
-
-            self.instances[nickname] = (host, port, auth)
+        self._instances = {}
+        self._clients = {}
+
+    def get_instances(self):
+      """Get the current instances to query for data.
+      """
+      instances = self.config['instances']
+      if callable(instances):
+        self.update_instance_cache(instances())
+      elif not self._instances:
+        self.update_instance_cache(instances)
+
+      return self._instances
+
+    def update_instance_cache(self, instance_list):
+      # configobj make str of single-element list, let's convert
+      if isinstance(instance_list, basestring):
+        instance_list = [instance_list]
+
+      # process original single redis instance
+      if len(instance_list) == 0:
+        host = self.config['host']
+        port = int(self.config['port'])
+        auth = self.config['auth']
+        if auth is not None:
+          instance_list.append('%s:%d/%s' % (host, port, auth))
+        else:
+          instance_list.append('%s:%d' % (host, port))
+
+      for instance in instance_list:
+
+        if '@' in instance:
+          (nickname, hostport) = instance.split('@', 1)
+        else:
+          nickname = None
+          hostport = instance
+
+        if '/' in hostport:
+          parts = hostport.split('/')
+          hostport = parts[0]
+          auth = parts[1]
+        else:
+          auth = None
+
+        if ':' in hostport:
+          if hostport[0] == ':':
+            host = self._DEFAULT_HOST
+            port = int(hostport[1:])
+          else:
+            parts = hostport.split(':')
+            host = parts[0]
+            port = int(parts[1])
+        else:
+          host = hostport
+          port = self._DEFAULT_PORT
+
+        if nickname is None:
+          nickname = str(port)
+
+        self._instances[nickname] = (host, port, auth)
```

Comment:
Is nickname only going to include the host, in some circumstances? Shouldn't it always be paired with the port to handle the case where there are multiple task instances running on the same host (at different ports)?

More generally, this is hard enough to follow that it might be worth adding example outputs in the docstring or more detailed comments inline.